### ISSUE TITLE
DataType conversion problems with to_dict()

### ISSUE THREAD

**Issue by davlee1972** [brunato/xmlschema#4]
Empty Dictionaries are being created for empty elements instead of None:

For some reason an empty tag like <NotPopulated></NotPopulated> is being converted into a empty dictionary when NotPopulated in the XSD is a xs:string.  Python should be converting these to None instead of {}. The JSON spec says it should be set to NotPopulated: null instead of NotPopulated: {}.

**Comment by brunato** [comment]
This issue should be fixed in the new version (0.9.7). The decode method of the XSD element now checks if the result_dict is empty and in this case return `None` if there isn't any text result. From lines 209-2016 of module xmlschema/components.py:

```
       if result_dict:
            if result:
                result_dict[text_key] = result
            yield result_dict
        elif result:
            yield result
        else:
            yield None
```

**Comment by davlee1972** [comment]
One more fix.. Looks like False is being converted to None:

       if result_dict:
            if result:
                result_dict[text_key] = result
            yield result_dict
        elif result or result is False:
            yield result
        elsif:
            yield None

**Comment by brunato** [comment]
True 👍

**Comment by davlee1972** [comment]
true and false are being converted to "true" and "false" if the are simple content. Is there a clean way to check if an element's complextype is simplecontent?

        <xs:complexType name="TYPE_ATTR_BOOLEAN">
                <xs:simpleContent>
                        <xs:extension base="xs:boolean">
                                <xs:attribute name="Type" type="xs:integer" use="optional"/>
                        </xs:extension>
                </xs:simpleContent>
        </xs:complexType>

Fix 1:

    if result_dict:
        if result or result is False:
            result_dict[text_key] = result
        yield result_dict
    elif result or result is False:
        yield result
    elsif:
        yield None

Fix 2 on Line 192:
if hasattr(self.type, 'content_type') and not self.type.has_extension() and not self.type.has_restriction():

**Comment by brunato** [comment]
I'm going to add a fix soon for this.

To check if an XsdElement is simple I will add the `is_simple()` method both to simple types base class (XsdSimpleType) and complex type class, so one can will simply check with `xsd_element.type.is_simple()`.

For Fix 2 do you refer to line 192 of components.py?

**Comment by davlee1972** [comment]
Looks like it is line 203 in components.py.. unicode_type("false") is being converted to a string "false".

Not sure if how the check if hasattr(self.type, 'content_type'): works without hacking in xsd_element.type.is_simple()

            if hasattr(self.type, 'content_type'):
                result = unicode_type(text)
            else:

**Comment by brunato** [comment]
This issue should be fixed in the new version (0.9.8), available also in pip.

Note:  the module `xmlschema/components.py` has been transformed into a sub-package (`components/`).
